src/browser-polyfill.js

```diff
@@ -132,7 +132,48 @@ if (typeof browser === "undefined") {
         }
 
         return new Promise((resolve, reject) => {
-          target[name](...args, makeCallback({resolve, reject}, metadata));
+          function callAPIWithNoCallback() {
+            try {
+              target[name](...args);
+              if (chrome.runtime.lastError) {
+                // NOTE: Ideally this should rejects if there are runtime issues with
+                // the API call on Chrome, unfortunately currently when pageAction.show/hide
+                // are called on Chrome with non existent tab ids, the `chrome.runtime.lastError`
+                // has been already logged on the console and it has been already set to undefined here,
+                // and so it doesn't seem to be currently possible to reject the promise
+                // with the chrome.runtime.lastError as the rejected value for the pageAction.show/hide.
+                reject(chrome.runtime.lastError);
+              } else {
+                // Update the API method metadata, so that the next API calls will not try to
+                // use the unsupported callback anymore.
+                metadata.fallbackToNoCallback = false;
+                metadata.noCallback = true;
+
+                resolve();
+              }
+            } catch (error) {
+              // Catch API parameters validation errors and rejects the promise.
+              reject(error);
+            }
+          }
+
+          if (metadata.fallbackToNoCallback) {
+            // This API method has currently no callback on Chrome, but it return a promise on Firefox,
+            // and so the polyfill try to pass a callback but fallback to not passing the callback
+            // if the first call have been rejected.
+            try {
+              target[name](...args, makeCallback({resolve, reject}, metadata));
+            } catch (cbError) {
+              console.warn(`${name} API method doesn't seem to support the callback parameter, ` +
```

Comment:
Under normal circumstances, this message will be displayed once.

This message will be displayed more than once if the extension passes the wrong arguments. Do you consider that to be OK? (personally I think that it is fine, since the code is simpler in this way; if you want to force the message to be displayed once, then it needs to be moved inside the `else` block above.

Why do we need log spam in the first place?

Reply:
I thought to emit at least a warning log mostly because it is not completely impossible that this runtime discovery can be fooled by some chrome behaviors (maybe a future chrome behavior) that the polyfill is not expecting, and it could completely hide the real error.

Emitting the warning every time we are actually falling back between the "with callback" to the "no callback" mode is fine from my current point of view, in the common scenarios it should only be produced once, and if it is produced more than once, then it could be a signal of something different between the behavior that it expect and the actual one.